cloudstate/cloudstate.go

```diff
@@ -0,0 +1,249 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cloudstateio/go-support/cloudstate/protocol"
+	"github.com/golang/protobuf/descriptor"
+	"github.com/golang/protobuf/proto"
+	filedescr "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+	"log"
+	"net"
+	"os"
+	"runtime"
+)
+
+const (
+	SupportLibraryVersion = "0.1.0"
+	SupportLibraryName    = "cloudstate-go-support"
+)
+
+// CloudState is an instance of a CloudState User Function
+type CloudState struct {
+	server                   *grpc.Server
+	entityDiscoveryResponder *EntityDiscoveryResponder
+	eventSourcedHandler      *EventSourcedHandler
+}
+
+// NewCloudState returns a new CloudState instance.
+func NewCloudState(options *Options) *CloudState {
+	cs := &CloudState{
+		server:                   grpc.NewServer(),
+		entityDiscoveryResponder: NewEntityDiscoveryResponder(options),
+		eventSourcedHandler:      NewEventSourcedHandler(),
+	}
+	protocol.RegisterEntityDiscoveryServer(cs.server, cs.entityDiscoveryResponder)
+	log.Println("RegisterEntityDiscoveryServer")
+	protocol.RegisterEventSourcedServer(cs.server, cs.eventSourcedHandler)
+	log.Println("RegisterEventSourcedServer")
+	return cs
+}
+
+// Options go get a CloudState instance configured.
+type Options struct {
+	ServiceName    string
+	ServiceVersion string
+}
+
+var NoOptions = Options{}
+
+// DescriptorConfig configures service and dependent descriptors.
+type DescriptorConfig struct {
+	Service        string
+	ServiceMsg     descriptor.Message
+	Domain         []string
+	DomainMessages []descriptor.Message
+}
+
+func (dc DescriptorConfig) AddDomainMessage(m descriptor.Message) DescriptorConfig {
+	dc.DomainMessages = append(dc.DomainMessages, m)
+	return dc
+}
+
+func (dc DescriptorConfig) AddDomainDescriptor(filename string) DescriptorConfig {
+	dc.Domain = append(dc.Domain, filename)
+	return dc
+}
+
+// Register registers an event sourced entity for CloudState.
+func (cs *CloudState) Register(ese *EventSourcedEntity, config DescriptorConfig) (err error) {
+	ese.registerOnce.Do(func() {
+		if err = ese.initZeroValue(); err != nil {
+			return
+		}
+		if err = cs.eventSourcedHandler.registerEntity(ese); err != nil {
+			return
+		}
+		if err = cs.entityDiscoveryResponder.registerEntity(ese, config); err != nil {
+			return
+		}
+	})
+	return
+}
+
+// Run runs the CloudState instance.
+func (cs *CloudState) Run() error {
+	host, ok := os.LookupEnv("HOST")
+	if !ok {
+		return fmt.Errorf("unable to get environment variable \"HOST\"")
+	}
+	port, ok := os.LookupEnv("PORT")
+	if !ok {
+		return fmt.Errorf("unable to get environment variable \"PORT\"")
+	}
```

Comment:
I was guided here by `io.cloudstate.javasupport.CloudStateRunner` and I don't know what CloudState here expects. It seems for me the API surface here is minimised deliberately but I have to check what was the intention by CloudState.

- [ ] check whats the story for HOST/PORT configuration for a CloudState instance.

Reply:
That's a good question. I raised this on gitter as well just now. I think a library should just have a single well defined API for applications to interact with. If we have an implicit API via something like environment variables, it can get quite confusing what's going on. I think only the application (whatever owns `main.go`) should decide how to interact with its environment - either via command line args or environment variables. This is typically handled by a library like cobra or kingpin, which can automatically merge environment variables and command line parameters.

Reply:
This might also be a golang distinction from Java? I think it's more typical to handle all your command line options/env inside main.go, versus loading these environment things elsewhere in the codebase.

Reply:
Whereas the user function support library acts as a library or framework, I had the impression to be more on the side of a framework where the API surface is quite limited or reduced so to say and user functions can be implemented without too much ceremony. Having it mentioned in the chat, James or Viktor might pick up on this too as it was not yet discussed yet explicitly.

As we have now these questions where it could be a distinction between Java and Go or any other implementation, it seems it raises perhaps a question that is not answered by the expectations CloudState would have off of a support library and should be defined by it.

Otherwise, I totally agree with your point of having the owner of main.go the sole authority to decide how environment or any other configuration that lets bootstrap the binary.